keps/sig-api-machinery/3903-unknown-version-interoperability-proxy/README.md

```diff
@@ -0,0 +1,837 @@
+upgrade or downgrade), not every apiserver can serve every resource at every
+version.
+
+To fix this, we will add a filter to the handler chain in the aggregator which
+proxies clients to an apiserver that is capable of handling their request.
+
+## Motivation
+
+When an upgrade or downgrade is performed on a cluster, for some period of time
+the apiservers are at differing versions and are able to serve different sets of
+built-in resources (different groups, versions, and resources are all possible).
+
+In an ideal world, clients would be able to know about the entire set of
+available resources and perform operations on those resources without regard to
+which apiserver they happened to connect to. Currently this is not the case.
+
+Today, these things potentially differ:
+* Resources available somewhere in the cluster
+* Resources known by a client (i.e. read from discovery from some apiserver)
+* Resources that can be actuated by a client
+
+This can have serious consequences, such as namespace deletion being blocked
+incorrectly or objects being garbage collected mistakenly.
+
+### Goals
+
+* Ensure discovery reports the same set of resources everywhere (not just group
+  versions, as it does today)
+* Proxy client traffic to an apiserver that can service the request
+* If traffic can't be proxied (e.g. if no network connection between
+  apiservers), return a 503 and not a 404 so clients can make the right
+  decision.
+
+### Non-Goals
+
+* Change cluster installation procedures (no new certs etc)
+* Lock particular clients to particular versions
+
+## Proposal
+
+We will use the existing `StorageVersion` API to figure out which group, versions,
+and resources an apiserver can serve.
+
+
+API server change:
+* A new handler is added to the stack:
+
+  - If the request is for a group/version/resource the apiserver doesn't have
+    locally (we can use the StorageVersion API), it will proxy the request to
+    one of the apiservers that is listed in the object. If an apiserver fails
+    to respond is not available, then we will return a 503 (there is a small
+    possibility of a race between the controller registering the apiserver
+    with the resources it can serve and receiving a request for a resource
+    that is not yet available on that apiserver).
+
+* Discovery merging.
+
+  - During upgrade or downgrade, it may be the case that no apiserver has a
+    complete list of available resources. To fix the problems mentioned, it's
+    necessary that discovery exactly matches the capability of the system. So,
+    we will use the storage version objects to reconstruct a merged discovery
+    document and serve that in all apiservers.
+
+### User Stories (Optional)
+
+#### Garbage Collector
+
+The garbage collector makes decisions about deleting objects when all
+referencing objects are deleted. A discovery gap / apiserver mismatch, as
+described above, could result in GC seeing a 404 and assuming an object has been
```

Comment:
If we make the 404 a 503, then this point is moot.

Reply:
I updated the goals and proposal section to make it clearer why this doesn't work.

Reply:
Yeah but even UVIP as currently stated doesn't really solve the problem, because in the case the apiserver is able to act on resources which end up disappearing in the newer version is kinda moot without the transition hooks which allow us to clean up our resource mess once the upgrade completes. And, in that same situation, the actions that the apiserver makes on those resources may not even make sense once the upgrade completes because those resources will essentially be orphaned. 

Therefore my contention is that 503ing here is not any less correct then letting the apiserver act on old resources which will then disappear once the upgrade completes **unless** we also have the ability to transition between resources being enabled/disabled.

Reply:
As stated I think this solves GC and namespace lifecycle controller, and doing less than this proposal would not (see all the paragraphs I added in the most recent commit).

It also enables a hypothetical clear-before-resource-removal controller, but that's not a story because it's not necessary to motivate any of the changes proposed here. GC and NLC already exercise the cases.